api/access/server.go

```diff
@@ -160,12 +160,97 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
 	return nil, errors.New("not implemented")
 }
 
-func (s *Server) GetTransaction(ctx context.Context, in *access.GetTransactionRequest) (*access.TransactionResponse, error) {
-	return nil, errors.New("not implemented")
+func (s *Server) GetTransaction(_ context.Context, in *access.GetTransactionRequest) (*access.TransactionResponse, error) {
+	id := flow.HashToID(in.Id)
+	tb, err := s.index.Transaction(id)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve transaction: %w", err)
+	}
+
+	resp := access.TransactionResponse{
+		Transaction: convert.TransactionToMessage(*tb),
+	}
+
+	return &resp, nil
 }
 
-func (s *Server) GetTransactionResult(ctx context.Context, in *access.GetTransactionRequest) (*access.TransactionResultResponse, error) {
-	return nil, errors.New("not implemented")
+func (s *Server) GetTransactionResult(_ context.Context, in *access.GetTransactionRequest) (*access.TransactionResultResponse, error) {
+	txID := flow.HashToID(in.Id)
+	tx, err := s.index.Transaction(txID)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve transaction: %w", err)
+	}
+
+	result, err := s.index.Result(txID)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve transaction result: %w", err)
+	}
+
+	// The code below is a simplified and limited implementation to derive the transaction status from
+	// https://github.com/onflow/flow-go/blob/v0.17.4/engine/access/rpc/backend/backend_transactions.go#L257-L328
+	statusCode := uint32(0)
+	if result.ErrorMessage == "" {
+		statusCode = 1
+	}
+
+	// First, we need to retrieve the last indexed height to access the latest seals.
+	lastHeight, err := s.index.Last()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve last height: %w", err)
+	}
+
+	sealIDs, err := s.index.SealsByHeight(lastHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve block seals: %w", err)
+	}
+
+	// We then go through all sealed blocks at the last height and look for the latest sealed height.
+	var sealedHeight uint64
+	for _, sealID := range sealIDs {
+		seal, err := s.index.Seal(sealID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve seal: %w", err)
+		}
+
+		h, err := s.index.HeightForBlock(seal.BlockID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve block height: %w", err)
+		}
+
+		if h > sealedHeight {
+			sealedHeight = h
+		}
+	}
```

Comment:
This seems unnecessarily complex, and it also doesn't work for a frequent edge case, where a block won't contain any seals. In the Flow Go code, we work around this, by indexing the last seal for each block when its finalized, even if that last seal was in a previous block. We probably need to do something similar.

Reply:
The best we can do is move this complexity around though, unless I'm missing smth. When building the index, to index the last sealed height, we need to look at all the seals that we got, lookup the height for their blockID and pick the highest one. This means that we have to do read operations in the index writer as well which I'm not fond of :/ Is there a better solution I'm missing here?

Currently in Flow they do something kind of weird:

```go
func (e *Engine) checkLastSealed(finalizedID flow.Identifier) error {
	// TODO: better to query seals from protocol state,
	// switch to state.Final().LastSealed() when available
	seal, err := e.sealsDB.ByBlockID(finalizedID)
	if err != nil {
		return fmt.Errorf("could not get the last sealed for the finalized block: %w", err)
	}

	blockID := seal.BlockID
	sealedCommit := seal.FinalState

	mycommit, err := e.execState.StateCommitmentByBlockID(e.unit.Ctx(), blockID)
	if errors.Is(err, storage.ErrNotFound) {
		// have not executed the sealed block yet
		// in other words, this can't detect execution fork, if the execution is behind
		// the sealing
		return nil
	}

	if err != nil {
		return fmt.Errorf("could not get my state commitment OnFinalizedBlock, blockID: %v", blockID)
	}

	if mycommit != sealedCommit {
		sealed, err := e.state.AtBlockID(blockID).Head()
		if err != nil {
			return fmt.Errorf("could not get sealed block when checkLastSealed: %v, err: %w", blockID, err)
		}

		return fmt.Errorf("execution result is different from the sealed result, height: %v, block_id: %v, sealed_commit: %x, my_commit: %x",
			sealed.Height, blockID, sealedCommit, mycommit)
	}

	return nil
}
```

And they do mention that they should switch and directly use the protocol state but unfortunately we can't take such shortcuts.

In another place (in the consensus nodes), this is what they do:

```go
	seal, err := c.seals.ByBlockID(finalizedBlockID)
	if err != nil {
		c.log.Fatal().Err(err).Msgf("could not retrieve seal for finalized block %s", finalizedBlockID)
	}
	lastSealed, err := c.headers.ByBlockID(seal.BlockID)
	if err != nil {
		c.log.Fatal().Err(err).Msgf("could not retrieve last sealed block %s", seal.BlockID)
	}

	// it's important to use atomic operation to make sure that we have correct ordering
	atomic.StoreUint64(&c.atomicLastSealedHeight, lastSealed.Height)
```

I see a few options but none of them seem great to me:

1. Add yet another index for SealsByBlockID and assume that any block that is at least at a height of 3 below the last indexed height is finalized. We can then do something like `lastSealedHeight := HeightForBlockID(SealByBlockID(BlockIDForHeight(LastHeight - 3)))` to get what we need.
2. When we index seals, read the height for the blocks they point to and store the highest one in a `LastSealed` height index (fairly simple, but means we do reads in the writer)
3. Keep the current complexity in the API rather than in the indexes

What makes the most sense to you guys? Am I missing smth obvious here? @awfm9 @Maelkum